### PR TITLE
Add scanWindow to optionally limit scan area

### PR DESCRIPTION
**Pull request by casvanluijtelaar**
Implements https://github.com/juliansteenbakker/mobile_scanner/issues/11 for IOS, Android & MacOS

### CODE REVIEW

**Comment by juliansteenbakker** [comment]
Thank you for your contribution! Can you please run `flutter format .` and push the changes?

**Comment by casvanluijtelaar** [comment]
just a heads up, I can confirm the new functionality on Android. But I can't seem to be able to can anything on IOS, using a scanWindow or not, and I'm not sure if that's because of something added in this pull. since Scanwindow code is completely skipped over if not provided.

**Comment by casvanluijtelaar** [comment]
Yeah there is still an issue with the cropping on ios. the rect is passed through correctly and a cropped image is returned from `cropSampleBuffer` just no QR code is being recognized through it. So I'm pretty sure I'm destroying the buffer data somewhere.

File: ios/Classes/SwiftMobileScannerPlugin.swift [hunk 1]
```diff
@@ -238,6 +240,23 @@ public class SwiftMobileScannerPlugin: NSObject, FlutterPlugin, FlutterStreamHan
...
+        let scanWindowData: Array? = argReader.intArray(key: "scanWindow")
+        if(scanWindowData != nil) {
+
+            let captureMetadataOutput = AVCaptureMetadataOutput()
+
+            captureMetadataOutput.rectOfInterest = CGRect(
+                    x: scanWindowData![0],
+                    y: scanWindowData![1],
+                    width: scanWindowData![2] - scanWindowData![0],
+                    height: scanWindowData![3] - scanWindowData![1])
```
Comment (casvanluijtelaar) [inline ios/Classes/SwiftMobileScannerPlugin.swift:254]: @juliansteenbakker I found a way better way to implement this, using the camera's `rectOfInterest` however I'm pretty sure this requires a CGRect mapped from 0 to 1

File: ios/Classes/SwiftMobileScannerPlugin.swift [hunk 2]
```diff
@@ -238,7 +240,25 @@ public class SwiftMobileScannerPlugin: NSObject, FlutterPlugin, FlutterStreamHan
...
         captureSession.startRunning()
+        
         let demensions = CMVideoFormatDescriptionGetDimensions(device.activeFormat.formatDescription)
+       
+
+         /// limit captureSession area of interest to the scanWindow if provided
+        let scanWindowData: Array? = argReader.intArray(key: "scanWindow")
+        if(scanWindowData != nil) {
+
+            let captureMetadataOutput = AVCaptureMetadataOutput()
```
Comment (casvanluijtelaar) [inline ios/Classes/SwiftMobileScannerPlugin.swift:251]: any way to integrate this with the existing `AVCaptureVideoDataOutput`

**Comment by juliansteenbakker** [comment]
Thank you for all your work! I will look into this to see if i can get AVCaptureMetadataOutput flow integrated.

**Comment by ayavn** [comment]
@casvanluijtelaar I had an app crash when using scanWindow.
```
java.lang.IllegalArgumentException: Image dimension, ByteBuffer size and format don't match. Please check if the ByteBuffer is in the decalred format.
```

Device: Samsung Galaxy A22. Android 12.

```
flutter doctor 

[✓] Flutter (Channel stable, 3.0.1, on macOS 11.6 20G165 darwin-arm, locale en)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.67.2)
[✓] Connected device (4 available)
[✓] HTTP Host Availability

• No issues found!
```

**Comment by casvanluijtelaar** [comment]
big change in approach, instead of actively cropping just compare the barcode locations to the frame.
the only thing that needs to happen now is make sure the Rect dimensions coming from flutter match the android and ios rects

**Comment by casvanluijtelaar** [comment]
I'd like some input on how to approach the Rect implementation, specifically because the screen size in flutter and the video size don't reflect each other 
```
flutter screen size: Size(428.0, 926.0)
video size: Size(3024.0, 4032.0)
```

from a flutter perspective I'd intuitively want to use the flutter screen size, so that a Rect for the camera cutout is a 1 to 1 match for a rect i'd perhaps want to overlay in a custom paint.

but that of course makes for a trickier conversion. I guess we could wait on the flutter side till we have the video width and height and then converting that to a screen size.

similar how it's done in the flutter widget itself :
image

**Comment by casvanluijtelaar** [comment]
shouldn't the `MobileScannerArguments` size be in flutter widget size too? that would make life easier. knowing the actual camera capture size seems pretty useless to me

**Comment by casvanluijtelaar** [comment]
and how would that work with a custom BoxFit

**Comment by casvanluijtelaar** [comment]
something like this:

```dart
  /// for boxfit.contain
  Size get actualSizeOfTextureInFlutterCoordinates {
    final screenWidth = window.physicalSize.shortestSide / window.devicePixelRatio;
    final screenHeight = window.physicalSize.longestSide / window.devicePixelRatio;

    final screenSize = Size(screenWidth, screenHeight);

    final ratio = screenSize.longestSide / screenSize.longestSide;

    return Size(size.width * ratio, size.height * ratio);
  }
```

**Comment by casvanluijtelaar** [comment]
the coordinate grids are extremely inconsistent.

- Oneplus 7 Pro
    - texture size: 1080, 1440
    - scanned image size: 640, 480
- Iphone 13 Pro Max
    - texture size: 3024, 4032
    - scanned image size: 1284, 1712
  
and then the flutter logical pixel size of the widget is different yet. how do we convert the incoming Rect so that  its consistent, and aligned, on all platforms.

**Comment by casvanluijtelaar** [comment]
I feel like I'm doing a terrible job explaining the issue:
image

any `Rect` defined in flutter will have it's coordinates relative to the widget size. But for the native barcode scanning, we need a rect relative to the image it scans. 

This is not an easy thing to do. first of all because depending on the `boxfit` the scan area could fall partly outside the texture, the texture couldn't have the same aspect ratio, ... etc. 

And on top of that, as mentioned earlier, texture size != actual image size used by Mlkit. so the coordinates would have to be scaled to match.

**Comment by casvanluijtelaar** [comment]
Progress, this might be a functional implementation on the flutter side to overcome all these issues:

```dart
  void calculateScanWindowRelativeToTexture(BoxFit fit, Rect scanWindow, Size textureSize, Size widgetSize) {

    /// map the texture size to get its new size after fitted to screen 
    final fittedTextureSize = applyBoxFit(fit, textureSize, widgetSize).destination;
    final minX = widgetSize.width / 2 - fittedTextureSize.width / 2;
    final minY = widgetSize.height / 2 - fittedTextureSize.height / 2;
    final textureWindow = Rect.fromLTWH(minX, minY, fittedTextureSize.width, fittedTextureSize.height);

    /// create a new scan window and with only the area of the rect intersecting the texture
    final scanWindowInTexture = scanWindow.intersect(textureWindow);

    /// update the scanWindow left and top to be relative to the texture not the widget
    final newLeft = scanWindowInTexture.left - minX;
    final newTop = scanWindowInTexture.top - minY;

    /// new scanWindow that is adapted to the boxfit and relative to the texture
    final scanWindowRelativeToTexture = Rect.fromLTWH(newLeft, newTop, scanWindowInTexture.left, scanWindowInTexture.top);

    /// get the scanWindow as a percentage of the texture
    final percentageLeft =  scanWindowRelativeToTexture.left / fittedTextureSize.width;
    final percentageRight = scanWindowRelativeToTexture.right / fittedTextureSize.width;
    final percentageTop = scanWindowRelativeToTexture.top / fittedTextureSize.height;
    final percentagebottom = scanWindowRelativeToTexture.bottom / fittedTextureSize.height;

    /// this rectangle can be send to native code and used to cut out a rectangle of the scan image
    final percentageRelativeRect = Rect.fromLTRB(percentageLeft, percentageRight, percentageTop, percentagebottom);
  }
```

this returns a Rect that contains percentage data of its position on the texture. now in native code we can use these percentages to reconstruct a scan area in native code. (assuming the texture aspect ratio is the same as the scan image aspect ratio)

**Comment by casvanluijtelaar** [comment]
Almost there, just a couple bugs in the previously mentioned method left

**Comment by juliansteenbakker** [comment]
Thank you for all your work and explanation! I've tested it on both android and iOS, and while i had no problems on iOS, i did have problems on Android. I tried it with 2 devices however it seems like the scanning area expands upwards of the RECT. Also, it seems quite hard to scan something in general on Android rather that on iOS. Do you have the same problems or is everything working for you on Android?

**Comment by casvanluijtelaar** [comment]
> Thank you for all your work and explanation! I've tested it on both android and iOS, and while i had no problems on iOS, i did have problems on Android. I tried it with 2 devices however it seems like the scanning area expands upwards of the RECT. Also, it seems quite hard to scan something in general on Android rather that on iOS. Do you have the same problems or is everything working for you on Android?

I can't replicate that with the example I added in the pull request, are you using a different setup?

**Comment by casvanluijtelaar** [comment]
about the slower scanning on Android I notice the same, but the Rect seems to work.

but the image that is scanned for android is (for me): 640 x 480
whilst on Iphone it is (for me): 1284 x 1712

if you for example have a full screen camera on android and you cut-out a centered rect, you're going to have an area of less than 100x100 pixels to find and recognize a scan code, Is there any reason the android image is this low quality?

**Comment by casvanluijtelaar** [comment]
can be changed apparently by doing `analysisBuilder.setTargetResolution`.

**Comment by casvanluijtelaar** [comment]
latest push should resolve the issues for android

**Comment by casvanluijtelaar** [comment]
We are noticing similar issues on android, you can seen one one android device it works perfectly whilst on another it doesn't seem to work at all. And at this point I'm at a loss what could be causing it.

https://user-images.githubusercontent.com/67225314/173818267-86f97069-c991-42f0-8ebb-12e5d91baf55.mp4
https://user-images.githubusercontent.com/67225314/173818241-cc717d5e-4120-4f2e-946e-c421be9f8fa9.mp4

**Comment by casvanluijtelaar** [comment]
current issue:

```kotlin
 // Build the analyzer to be passed on to MLKit
val analysisBuilder = ImageAnalysis.Builder()
      .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
      .setTargetResolution(/* set resolution the same as preview, however preview size isn't defined yet */)
```

how do we get the preview size at this point? mobilescanner.kt:224

**Comment by casvanluijtelaar** [comment]
issue:

```kotlin
  val previewBuilder = Preview.Builder()
  if (ratio != null) { previewBuilder.setTargetAspectRatio(ratio) }
  
  preview = previewBuilder.build().apply { setSurfaceProvider(surfaceProvider) }

  val analysisBuilder = ImageAnalysis.Builder()
          .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
          /// preview!!.resolutionInfo?.resolution is still null at this point so we can't use it.
          .setTargetResolution(preview!!.resolutionInfo?.resolution ?: Size(1920, 1080))


  if (ratio != null) { analysisBuilder.setTargetAspectRatio(ratio) }
  val analysis = analysisBuilder.build().apply { setAnalyzer(executor, analyzer) }

  val selector = if (facing == 0) CameraSelector.DEFAULT_FRONT_CAMERA else CameraSelector.DEFAULT_BACK_CAMERA

  /// preview resolution is defined at the same time as the analyzer is set, so there's no way to have them
  /// match resolutions up from. at this point the above code will fall back to Size(1920, 1080) which can in no way 
  /// guarantee the same resolution and aspect ratio and therefore matching Rects. IOS matches resolutions
  /// automatically so this isn't an issue
  camera = cameraProvider!!.bindToLifecycle(activity as LifecycleOwner, selector, preview, analysis)

  val analysisSize = analysis.resolutionInfo?.resolution ?: Size(0, 0)
  val previewSize = preview!!.resolutionInfo?.resolution ?: Size(0, 0)
  
  Log.i("LOG", "Analyzer: $analysisSize")
  Log.i("LOG", "Preview: $previewSize")

```

**Comment by juliansteenbakker** [comment]
Sorry again for the late response and thanks for all the hard work! I'll try and see if i can get some other devices to test it on and see if this maybe is an issue with the camera2 library

**Comment by casvanluijtelaar** [comment]
I just started working on this again today @juliansteenbakker ;), found the root of the issue, well multiple. 

- ImageAnalyzer allows you only either set resolution or aspectRatio, and since aspectRatio is required it might take a very low resolution. 
- the camera rotation is currently pretty messed up, ImagAnalyzer's ImageProxy is flipped on its side, even when setting its targetRotation to match the preview.

have a look at my branch for latest updates

**Comment by juliansteenbakker** [comment]
Nice! So i merged your changes in a local branch so i can publish it as a beta. If you want you can create a new pull request and merge it against feature/scan-window instead of master. After that i will publish it as a beta release!

**Comment by mehdiwaysi** [comment]
This feature is very critical, please add it

**Comment by vincnet500** [comment]
Could it be possible to release the beta version ?

**Comment by weili1210** [comment]
Hi, just want to ask is this feature available on beta version or not?

**Comment by ArslanAsghar123** [comment]
Hi all, is there any possible way to add texts inside the overlay?

**Pull request merged**